File: algorithms/python/union_find.py
class UnionFind:
    def __init__(self, n: int):
        self.n = n
        self.parent = list(range(n))
        self.rank = [1] * n

    def find(self, i: int) -> int:
        # Path Compression
        if self.parent[i] != i:
            self.parent[i] = self.find(self.parent[i])
        return self.parent[i]

    # Union by Rank
    def merge(self, i: int, j: int) -> None:
        pi, pj = self.find(i), self.find(j)
        if self.rank[pi] < self.rank[pj]:
            pi, pj = pj, pi

        if self.rank[pi] == self.rank[pj]:
            self.rank[pi] += 1
        self.parent[pj] = pi

def test():
    uf = UnionFind(n=5)
    uf.merge(0, 1)
    assert uf.rank[0] == 2
    assert uf.find(0) == 0
    assert uf.find(1) == 0

    uf.merge(2, 3)
    assert uf.find(3) == 2

    uf.merge(1, 3)
    assert uf.find(2) == 0
    assert uf.find(3) == 0
    assert uf.rank[0] == 3

    print("Test Passed!")
    return

if __name__ == "__main__":
    test()

File: algorithms/cpp/union-find.cpp
using namespace std;

#include <vector>

class UnionFind {
  int n = 0;
  vector<int> parent, rank;

  UnionFind(int _n) {
    n = _n;
    for (int i = 0; i < n; i++) {
      parent[i] = i;
      rank[i] = 1;
    }
  }

  int find(int v) {
    if (v == parent[v]) {
      return v;
    }
    parent[v] = find(parent[v]);
    return parent[v];
  }

  void merge(int u, int v) {
    u = find(u), v = find(v);
    if (u == v) {
      return;
    }

    if (rank[u] > rank[v]) {
      swap(u, v);
    }
    parent[u] = v;
    if (rank[u] == rank[v]) {
      rank[v]++;
    }
  }
};
